lua/nvim-tree/actions/fs/create-file.lua

```diff
@@ -41,6 +37,73 @@ local function get_num_nodes(iter)
   return i
 end
 
+local function create_file(new_file_path)
+  -- create a folder for each path element if the folder does not exist
+  -- if the answer ends with a /, create a file for the last path element
+  local is_last_path_file = not new_file_path:match(utils.path_separator .. "$")
+  local path_to_create = ""
+  local idx = 0
+
+  local num_nodes = get_num_nodes(utils.path_split(utils.path_remove_trailing(new_file_path)))
+  local is_error = false
+  for path in utils.path_split(new_file_path) do
+    idx = idx + 1
+    local p = utils.path_remove_trailing(path)
+    if #path_to_create == 0 and utils.is_windows then
+      path_to_create = utils.path_join { p, path_to_create }
+    else
+      path_to_create = utils.path_join { path_to_create, p }
+    end
+    if is_last_path_file and idx == num_nodes then
+      if M.enable_async then
+        async.schedule()
+      end
+      if utils.file_exists(new_file_path) then
+        local prompt_select = "Overwrite " .. new_file_path .. " ?"
+        local prompt_input = prompt_select .. " y/n: "
+        if M.enable_async then
+          local item_short = async.call(lib.prompt, prompt_input, prompt_select, { "y", "n" }, { "Yes", "No" })
```

Comment:
Do we need to make the prompt async? Only `create_and_notify` etc. needs to be async.

Reply:
Yes, this is mandatory in async version. In fact, `lib.prompt` itself is already async because of usage of callback, the only difference is that the async context of `lib.prompt` is not controlled. If we use callback instead of wrapping with `async.call`, `create_and_notify` may not be called in the same async context as where `async.exec` begins and the immediate following `async.call` in `create_and_notify` will throw error.

You can test it will override `vim.ui.input` with https://github.com/stevearc/dressing.nvim or even a simple delayer:
```lua
local original = vim.ui.input
vim.ui.input = function(opts, on_confirm)
  original(opts, function(input)
     vim.defer_fn(function() on_confirm(input) end, 0)
  end)
end
```

Reply:
In principle, all callback-style functions should be wrapped to make the async context consistent.

Reply:
Unfortunately I still don't understand. Yes, the vim.ui.select calls themselves are asynchronous, however the callbacks appear to be executed on the main thread / context.

```lua
function M.prompt(prompt_input, prompt_select, items_short, items_long, callback)
  ...
  if M.select_prompts then
    vim.ui.select(items_short, { prompt = prompt_select, format_item = format_item }, function(item_short)
      log.line("dev", "prompt select in_fast_event=%s is_thread=%s", vim.in_fast_event(), vim.is_thread())
      callback(item_short)
    end)
  else
    vim.ui.input({ prompt = prompt_input }, function(item_short)
      log.line("dev", "prompt input in_fast_event=%s is_thread=%s", vim.in_fast_event(), vim.is_thread())
      callback(item_short)
    end)
  end
end
```

```
[2023-01-23 10:29:44] [dev] prompt input in_fast_event=false is_thread=false
```

Reply:
Yes, because it might be called in main thread (not in async thread), we need to use `async.call` to transfer it to async thread, otherwise `create_and_notify` will throw error because `async.call` uses `coroutine.yield` under the hood.

Reply:
All our vim.ui.input and vim.ui.select calls are from the main thread. If we made these calls from an async thread it would be an issue, however we have no need to do that.

Please revert the changes to the inputs. This change is about making the create and rename file system operations async, not the prompts. We can make a future change to allow async input calls if it is ever needed.

Reply:
I'm sorry that it is required to call the prompt callback in async thread, otherwise we have to use `async.exec` before every calls to following async code in the callback and it doesn't deserve IMO. And additionally, in the current implementation `vim.ui.*` is actually called in the main thread, only the callback is called in coroutine. I think technically there would be no issue here.

Reply:
Let's just focus on `rename` for now as it is less complex than `create`.

Outcome: `M.rename` `vim.loop.fs_rename` call is asynchronous
Condition: `async.exec` must be called from the main thread
Problem: `M.rename` is executed from a `vim.ui.input` callback which may not be on the main thread

Proposal: call `M.rename` after the input, outside of the `vim.ui.input` callback

Reply:
`async.exec` is not required to be called from main thread. It's possible to create new coroutine in another coroutine. The coroutine (or thread, here) is just stackful statemachine and could be created anywhere.

Would you mind I make a simple guidance for understanding of current basic async implementation? I've seen that there are plenty of misunderstandings between us and possibly with future contributors.

Reply:
Yes please. A detailed explaination would be gratefully appreciated.

The outcome of this PR:
* file system operations for create / rename are asynchronous
* code changes are minimal and clear
* asynchronous operations are localised i.e. not spread over multiple callbacks / functions

Unless there is a very good reason, inputs should not be asynchronous. This adds unnecessary complexity.

Reply:
Sorry for that I'm gonna to be off-topic, I think the referred async input codepath would not be ever reached, as we already checked the existence of file before `async.exec`: https://github.com/nvim-tree/nvim-tree.lua/blob/c71f47a8fa955d985018ae607b10cc741b79b965/lua/nvim-tree/actions/fs/create-file.lua#L136-L142. 

Do you think it's safe to remove this? I just cp from the original implementation. There is no other place with async call to  user input.

Reply:
I see... that's a good catch!

It does look like this `file_exits` will fail and exit early. The `local function create_file(new_file_path)` call to `file_exists` will always pass and the `"Overwrite " .. new_file_path .. " ?"` prompt cannot occur.

This is existing behaviour: we can safely remove the `file_exists` check in `create_file`.

Please raise a separate PR to remove this dead code path. We will merge that first, keeping our changes small and atomic and easily revertable.

Reply:
I found that I referred the wrong variable, `new_file_path` -> `path_to_create`, sorry about that. However, the truth that this is a dead code remains; and I think we should early return instead of prompting for overwrite for already existing file.